jupyterhub/schema.yaml

```
@@ -583,11 +583,86 @@ properties:
+              performance penalty.
+              ```
+            properties:
+              cloudMetadataServer:
+                type: boolean
+                description: |
+                  Defaults to `false` for singleuser servers, but to `true` for
+                  all other network policies.
+
+                  When enabled this rule allows the respective pod(s) to
+                  establish outbound connections to the cloud metadata server.
+
+                  Note that the `nonPrivateIPs` rule is allowing all non Private
+                  IP ranges but makes an exception for the cloud metadata
+                  server, leaving this as the definitive configuration to allow
+                  access to the cloud metadata server.
+              dnsPortsPrivateIPs:
+                type: boolean
+                description: |
+                  Defaults to `true` for all network policies.
+
+                  When enabled this rule allows the respective pod(s) to
+                  establish outbound connections to private IPs via port 53.
+
+                  Note that we can't reliably identify the k8s internal DNS
+                  server due to variations between k8s clusters. Due to that,
+                  this rule which is critical for core functionality, can be
+                  disabled for a more refined custom rule.
+              nonPrivateIPs:
+                type: boolean
+                description: |
+                  Defaults to `true` for all network policies.
+
+                  When enabled this rule allows the respective pod(s) to
+                  establish outbound connections to the non-private IP ranges
+                  with the exception of the cloud metadata server. This means
+                  respective pod(s) can establish connections to the internet
+                  but not (say) an unsecured prometheus server running in the
+                  same cluster.
+              privateIPs:
+                type: boolean
+                description: |
+                  Defaults to `false` for singleuser servers, but to `true` for
+                  all other network policies.
+
+                  Private IPs refer to the IP ranges `10.0.0.0/8`,
+                  `172.16.0.0/12`, `192.168.0.0/16`.
+
+                  When enabled this rule allows the respective pod(s) to
+                  establish outbound connections to the internal k8s cluster.
+                  This means users can access the internet but not (say) an
+                  unsecured prometheus server running in the same cluster.
+
+                  Since not all workloads in the k8s cluster may have
+                  NetworkPolicies setup to restrict their incoming connections,
+                  having this set to false can be a good defence against
+                  malicious intent from someone in control of software in these
+                  pods.
```

```suggestion
                  pods.
                  
                  To improve security we recommend only allowing the private IP CIDRs required
                  <TODO: insert example>
                  or namespaces
                  <TODO: insert example>
                  instead of enabling this.
```
Alternatively the example could go in to the docs https://zero-to-jupyterhub.readthedocs.io/en/latest/administrator/security.html#kubernetes-network-policies instead of embedding it here?

---

This is the description of a boolean value, toggling the inclusion of a egress rule alongside other rules. I don't think an example on how to configure something else belongs here if this would be for example disabled.

I'll see if I can push on and update the docs to include details about this new configuration in general and some ideas on recommendations.